spring-boot/proxy/src/test/java/security/ApiKeyAuthTokenVerifierTest.java

```diff
@@ -0,0 +1,189 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+import eu.coatrack.api.ApiKey;
+import eu.coatrack.api.ServiceApi;
+import eu.coatrack.proxy.security.*;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.springframework.security.web.authentication.session.SessionAuthenticationException;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class ApiKeyAuthTokenVerifierTest {
+
+    private final String someValidApiKeyValue = "ca716b82-745c-4f6d-a38b-ff8fe140ffd1";
+    private ApiKeyAuthToken apiKeyAuthToken;
+    private ApiKeyAuthTokenVerifier apiKeyAuthTokenVerifier;
+    private ApiKey apiKey;
+
+    private LocalApiKeyManager localApiKeyManagerMock;
+    private ApiKeyFetcher apiKeyFetcherMock;
+    private ApiKeyVerifier apiKeyVerifierMock;
+
+    @BeforeEach
+    public void setup(){
+        apiKey = createSampleApiKeyForTesting();
+
+        apiKeyAuthTokenVerifier = createApiKeyAuthTokenVerifier();
+
+        // Create an auth token for a valid api key without any granted authorities.
+        apiKeyAuthToken = new ApiKeyAuthToken(someValidApiKeyValue, null);
+        apiKeyAuthToken.setAuthenticated(false);
+    }
+
+    private ApiKey createSampleApiKeyForTesting() {
+        ServiceApi serviceApi = new ServiceApi();
+        serviceApi.setUriIdentifier("weather-data-service");
+
+        ApiKey localApiKey = new ApiKey();
+        localApiKey.setKeyValue(someValidApiKeyValue);
+        localApiKey.setServiceApi(serviceApi);
+
+        return localApiKey;
+    }
+
+    private ApiKeyAuthTokenVerifier createApiKeyAuthTokenVerifier() {
+        localApiKeyManagerMock = mock(LocalApiKeyManager.class);
+        apiKeyFetcherMock = mock(ApiKeyFetcher.class);
+        apiKeyVerifierMock = mock(ApiKeyVerifier.class);
+
+        ApiKeyAuthTokenVerifier localApiKeyAuthTokenVerifier = new ApiKeyAuthTokenVerifier(
+                localApiKeyManagerMock,
+                apiKeyFetcherMock,
+                apiKeyVerifierMock
+        );
+
+        return localApiKeyAuthTokenVerifier;
+    }
+
+    @Test
+    public void nullArgumentShouldCauseException(){
+        assertThrows(SessionAuthenticationException.class, () -> apiKeyAuthTokenVerifier.authenticate(null));
+    }
+
+    @Test
+    public void nullCredentialsInAuthTokenShouldCauseException(){
+        ApiKeyAuthToken token = new ApiKeyAuthToken(null, null);
+
+        assertThrows(SessionAuthenticationException.class, () -> apiKeyAuthTokenVerifier.authenticate(token));
```

Comment:
This should be a 
https://docs.spring.io/spring-security/site/docs/4.2.15.RELEASE/apidocs/org/springframework/security/authentication/BadCredentialsException.html

Reply:
See comment above.